File: Cracking The Coding Inverview - Book/01-arrays-and-strings/01-02-check-permutation/checkPermutation2.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

bool is_permutation(char* w1, char* w2)
{
    int letters[128] = { 0 };

    for (int i = 0; i < strlen(w1); i++)
    {
        letters[w1[i]]++;
    }

    for (int i = 0; i < strlen(w2); i++)
    {
        letters[w2[i]]--;
        if (letters[w2[i]] < 0)
        {
            return false;
        }
    }

    return true;
}

int main(void)
{
    if (is_permutation("renzo", "ozner") == true)
    {
        printf("true\n");
    }
    else
    {
        printf("false\n");
    }
}

File: Cracking The Coding Inverview - Book/01-arrays-and-strings/01-02-check-permutation/CheckPermutation2.java
// Check if the two words have the same character counts
 
public class CheckPermutation2 {

    static boolean isPermutation(String w1, String w2) {

        int[] letters = new int[128];

        // Is it better to convert the string to an array and iterate
        // the array, than to iterate the string directly??
        char[] arr1 = w1.toCharArray();
        for (char c : arr1) {
            letters[c]++;
        }

        for (int i = 0; i < w2.length(); i++) {
            int c = (int) w2.charAt(i);
            letters[c]--;
            if (letters[c] < 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {

        System.out.println(isPermutation("renzo", "ozner"));
    }
}
